packages/components/src/DataList/DataList.utils.tsx

```diff
@@ -32,6 +34,23 @@ export function getCompoundComponent<T>(
   return isValidElement<T>(element) ? element : undefined;
 }
 
+/**
+ * Return all instances child component that matches the `type` provided
+ */
+export function getCompoundComponents<T>(
+  children: ReactElement | ReactElement[],
+  type: ReactElement<T>["type"],
+): ReactElement<T>[] {
+  const childrenArray = Children.toArray(children);
+  const elements = childrenArray.filter(
+    (child): child is ReactElement<T> =>
+      isValidElement<T>(child) && child.type === type,
+  );
+
+  // Comply with the return type without casting it
+  return elements || [];
```

Comment:
filter returns an empty array if it finds nothing so you actually don't need this

```suggestion
  return elements;
```